File: java/explain_java/chapter13/Shape.java
// 图形类群

// -- 图形
abstract class Shape {
	abstract void draw();	// 绘图(抽象方法)
}

// -- 点
class Point extends Shape {
	Point() { }

	void draw() {
		System.out.println("+");
	}
}

// -- 长方形
class Rectangle extends Shape {
	private int width;
	private int height;

	Rectangle(int width, int height) {
		this.width = width;
		this.height = height;
	}

	void draw() {
		for (int i = 0; i < height; i++) { 
			for (int j = 0; j < width; j++) 
				System.out.print("*");
			System.out.println();
		}
	}
}

File: kotlin/kotlin_from_entry_to_advanced_combat/src/com/easykotlin/test/chapter4/Shape.kt
package com.easykotlin.test.chapter4

abstract class Shape  // 声明抽象类Shape

class Rectangle: Shape() // 继承类的语法是使用冒号“:”,父类需要在这里使用构造函数进行初始化

class Circle: Shape() // 

class Triangle: Shape()



fun main(args: Array<String>) {
    // val s = Shape() // 编译不通过！不能实例化抽象类
    val r = Rectangle()
    println(r is Shape)
}